### ISSUE TITLE
Terraform scripts for cloud cluster deployments

### ISSUE THREAD

**Issue by jhamman** [pangeo-data/pangeo#384]
It will be useful to have a set of terraform scripts to ease the deployment of cloud clusters. Currently we have shell scripts that look like this:

```bash
ZONE='us-central1-b'
NUM_NODES=2
MIN_WORKER_NODES=0
MAX_WORKER_NODES=10
CLUSTER_NAME='pangeo-jhub'

# create cluster on GCP
gcloud config set project $PROJECTID
gcloud container clusters create $CLUSTER_NAME \
    --num-nodes=$NUM_NODES \
    --disk-type=pd-ssd \
    --machine-type=n1-standard-2 \
    --zone=$ZONE \
    --no-enable-legacy-authorization

# Core pool
gcloud container node-pools create core-pool \
    --zone=$ZONE \
    --cluster=$CLUSTER_NAME \
    --machine-type=n1-highmem-4 \
    --preemptible \
    --enable-autoscaling \
    --num-nodes=$MIN_WORKER_NODES \
    --max-nodes=$MAX_WORKER_NODES \
    --min-nodes=$MIN_WORKER_NODES

# standard worker pool
gcloud beta container node-pools create worker-pool \
    --zone=$ZONE \
    --cluster=$CLUSTER_NAME \
    --machine-type=n1-standard-4 \
    --preemptible \
    --enable-autoscaling \
    --num-nodes=$MIN_WORKER_NODES \
    --max-nodes=$MAX_WORKER_NODES \
    --min-nodes=$MIN_WORKER_NODES \
    --disk-type=pd-ssd \
    --local-ssd-count 1

# big-mem worker pool
gcloud beta container node-pools create big-worker-pool \
    --zone=$ZONE \
    --cluster=$CLUSTER_NAME \
    --machine-type=n1-highmem-4 \
    --preemptible \
    --enable-autoscaling \
    --num-nodes=$MIN_WORKER_NODES \
    --max-nodes=$MAX_WORKER_NODES \
    --min-nodes=$MIN_WORKER_NODES \
    --disk-type=pd-ssd \
    --local-ssd-count 1

gcloud container clusters get-credentials $CLUSTER_NAME --zone=$ZONE --project $PROJECTID
```

Then we have a bunch of boilerplate commands:

```bash
# set up kubernetes
kubectl create clusterrolebinding cluster-admin-binding --clusterrole=cluster-admin --user=$EMAIL
kubectl create serviceaccount tiller --namespace=kube-system
kubectl create clusterrolebinding tiller --clusterrole=cluster-admin --serviceaccount=kube-system:tiller
helm init --service-account tiller
#helm init --wait --service-account tiller
kubectl --namespace=kube-system patch deployment tiller-deploy --type=json --patch='[{"op": "add", "path": "/spec/template/spec/containers/0/command", "value": ["/tiller", "--listen=localhost:44134"]}]'

# get helm repositories
helm repo add jupyterhub https://jupyterhub.github.io/helm-chart/
helm repo update

# install jupyterhub on the cluster
echo "Installing jupyterhub."
helm install jupyterhub/binder --version=$VERSION --name=jupyter --namespace=jupyter -f binder_config.yaml -f binder_secret.yaml
```

The top section changes based on the cloud provider while the bottom section is basically fixed. Having terraform scripts for each section would greatly simplify the deployment process and make it easier to duplicate/reproduce our clusters. 

I know some folks at UW are interested in helping in this area (@amanda-tan, @robfatland, @jonahjoughin, @aaarendt). @dsludwig may also have something to add here.

**Comment by dsludwig** [comment]
I took a shot at this, but ran into this issue pretty quickly: https://github.com/terraform-providers/terraform-provider-google/issues/2022

It creates the cluster and node pools, but it seems to leave things in a dirty state. I'll create a repository for the terraform manifests I've generated so far.

I believe https://github.com/yuvipanda/hubploy can perform the latter half (from `helm repo add` on).

**Comment by jhamman** [comment]
Thanks @dsludwig. Interestingly, I was getting some similar api errors using the gcloud commands above today too.

**Comment by guillaumeeb** [comment]
As already mentionned before, really interested about this possibility.

@jhamman where does your shell script commands come from? It does look different from what is described in [Cloud setup guide](https://pangeo.readthedocs.io/en/latest/setup_guides/cloud.html):
- you're creating several node pools, with core-pool vs worker-pools differentiation.
- you're installing jupyterhub/binder helm repo chart of pangeo one.

Did I miss something and the deployment procedure has been modified somewhere?

**Comment by jhamman** [comment]
@guillaumeeb - the helm install/upgrade line is not particularly important here. We have been experimenting with multiple node pools and the block of code I posted above reflects what I'm using for binder.pangeo.io. Eventually, I suspect we'll move some of this to the pangeo docs but in the meantime, this was just meant to illustrate functionality.

**Comment by guillaumeeb** [comment]
> I took a shot at this, but ran into this issue pretty quickly: terraform-providers/terraform-provider-google#2022

Same issue currently with GKE, even without terraform.

I'm currently using 
````
gcloud container node-pools create worker-pool --zone=$ZONE --cluster=$CLUSTER_NAME \
    --machine-type=$WORKER_MACHINE_TYPE --preemptible --num-nodes=$MIN_WORKER_NODES
gcloud container clusters update $CLUSTER_NAME --zone=$ZONE --node-pool=worker-pool \
    --enable-autoscaling --max-nodes=$MAX_WORKER_NODES --min-nodes=$MIN_WORKER_NODES
````
for my tests instead of
````
gcloud container node-pools create worker-pool --zone=$ZONE --cluster=$CLUSTER_NAME \
   --machine-type=$WORKER_MACHINE_TYPE --preemptible --enable-autoscaling \
   --num-nodes=$MIN_WORKER_NODES --max-nodes=$MAX_WORKER_NODES --min-nodes=$MIN_WORKER_NODES
````

**Comment by dsludwig** [comment]
My first cut is here: https://github.com/dsludwig/pangeo-terraform

It looks like Google support is trying to resolve the underlying issue that I was encountering with autoscaling here: https://github.com/terraform-providers/terraform-provider-google/issues/2022, so when that is solved, I'll pick it back up again.

**Comment by guillaumeeb** [comment]
Thanks @dsludwig. https://github.com/terraform-providers/terraform-provider-google/issues/2022 has been fixed earlier today, I tried your terraform config and it worked like a charm for cluster and pools creation.

I can see you've already began working with kubernetes commands integration into the TF config files. Some open points on my side:

- We should also use Terraform to ensure we have a public IP. We should use the public IP then to configure our yaml files.
- I'm not sure if all can be done from terraform for the helm part, even creating tiller service. I did not take a close look to https://github.com/yuvipanda/hubploy, but I'm not sure this can be run manually?
- We should probably move the work done by @dsludwig inside pangeo namespace. Should it be in a separate repo (pangeo-data/pangeo-terraform), or inside this one (somewhere here: https://github.com/pangeo-data/pangeo/tree/master/gce)?

Ideally, as @jhamman said above, we should have two scripts:

1. First one is responsible for creating Kubernetes cluster with a given cloud provider, Terraform is perfect for this.
2. Second one needs to initialize helm on Kubernetes and deploy the pangeo chart. Terraform wont be able to do everything, so we must probably look at other solutions like hubploy, maybe ansible? And I'm not sure what to do with https://github.com/informatics-lab/terraform-kubernetes.

Boundaries between the two parts are not that clear, maybe the first script can configure some tiller things...
